src/persist-txn/src/txn_read.rs

```diff
@@ -55,6 +55,10 @@ pub struct DataSnapshot<T> {
     pub(crate) as_of: T,
     /// An upper bound on the times known to be empty of unapplied writes via txns.
     pub(crate) empty_to: T,
+    /// The latest write > empty_to. None if there are no writes via txns > empty_to.
+    pub(crate) read_to: Option<T>,
+    /// An upper bound on the times know to be empty of writes via txn.
+    pub(crate) logical_progress: T,
```

Comment:
Maybe what we actually want here is the current physical and logical upper of the shard. The logical upper is always `cache.logical_progress`. The physical upper is `read_to.unwrap_or(empty_to)` if the shard is currently registered and `cache.logical_progress` if the shard is not currently registered.